types.go

```diff
@@ -0,0 +1,296 @@
+	RemoveFailedServer(*Server) error
+}
+
+type RaftChanges struct {
+	Promotions []raft.ServerID
+	Demotions  []raft.ServerID
+	Leader     raft.ServerID
+}
+
+type FailedServers struct {
+	// StaleNonVoters are the ids of those server in the raft configuration as non-voters
+	// that are not present in the delegates view of what servers should be available
+	StaleNonVoters []raft.ServerID
+	// StaleVoters are the ids of those servers in the raft configuration as voters that
+	// are not present in the delegates view of what servers should be available
+	StaleVoters []raft.ServerID
+
+	// FailedNonVoters are the servers without voting rights in the cluster that the
+	// delegate has indicated are in a failed state
+	FailedNonVoters []*Server
+	// FailedVoters are the servers without voting rights in the cluster that the
+	// delegate has indicated are in a failed state
+	FailedVoters []*Server
+}
+
+// Promoter is an interface to provide promotion/demotion algorithms to the core autopilot type.
+// The BasicPromoter satisfies this interface and will promote any stable servers but other
+// algorithms could be implemented. The implementation of these methods shouldn't "block".
+// While they are synchronous autopilot expects the algorithms to not make any network
+// or other requests which way cause an indefinite amount of waiting to occur.
+//
+// Note that all parameters passed to these functions should be considered read-only and
+// their modification could result in undefined behavior of the core autopilot routines
+// including potential crashes.
+type Promoter interface {
+	// GetServerExt returns some object that should be stored in the Ext field of the Server
+	// This value will not be used by the code in this repo but may be used by the other
+	// Promoter methods and the application utilizing autopilot. If the value returned is
+	// nil the extended state will not be updated.
+	GetServerExt(*Config, *ServerState) interface{}
+
+	// GetStateExt returns some object that should be stored in the Ext field of the State
+	// This value will not be used by the code in this repo but may be used by the other
+	// Promoter methods and the application utilizing autopilot. If the value returned is
+	// nil the extended state will not be updated.
+	GetStateExt(*Config, *State) interface{}
+
+	// GetNodeTypes returns a map of ServerID to NodeType for all the servers which
+	// should have their NodeType field updated
+	GetNodeTypes(*Config, *State) map[raft.ServerID]NodeType
+
+	// CalculatePromotionsAndDemotions
+	CalculatePromotionsAndDemotions(*Config, *State) RaftChanges
+
+	// CalculateFailedServerRemovals takes in the current state and structure outlining all the
+	// failed/stale servers and will return those failed servers which the promoter thinks
+	// should be allowed to be removed.
+	CalculateFailedServerRemovals(*Config, *State, *FailedServers) *FailedServers
```

Comment:
I was surprised to see `Config` passed to all of these. The config comes from the caller, so the caller could be responsible for creating a `Promoter` which has accept to either the same config, or the same `GetConfig()` method. Passing the config into `Autopilot` just to have it passed back out to the `Promoter` which came from the same place can make it difficult to understand how to use this package.

Reply:
The way its supposed to work is that Autopilot retrieves the config from the Delegate (which given your other comment probably could use a better name) and then uses the Promoter type for doing most of the calculations about promotions and demotions.

It is structured this way to allow for pluggable promotion algorithms. In Consul's case, and I would argue this is how it should be for all promoters, the Promoter and Delegate are completely independent and the Promoter has no access to the Config.

We could grant it access through other methods that it could use to retrieve the config but that would present one huge problem. During a single round of promotion/demotions we retrieve the config from the delegate exactly once at the beginning and that is the configuration we use for all determinations of promotions and demotions. If we made it so that the promoter had to go and fetch the config then potentially each call to a promoter method would have to fetch the config and potentially would be doing calculations with a different config than the previous call. This would end up with a potentially inconsistent state. So to prevent all of that we must pass in the Config and the Promoter must use that configuration and not a different one even if it could retrieve it directly.